Example: Debugging an Application with mirrord (#32)
metalbear-co/mirrord.dev: content/docs/overview/real-world-example.md

description: "Debugging an Application with mirrord .feat NextJS"
date: 2021-09-13T10:50:00+01:00
lastmod: 2021-09-13T10:50:00+01:00
draft: false
images: []
menu:
  docs:
    parent: "overview"
weight: 120
toc: true
---

## Setup

### Prerequisites 

* [mirrord](https://mirrord.dev/)
* Available Kubernetes Cluster
* [NodeJS](https://nodejs.org/en/) + [Yarn](https://www.npmjs.com/package/yarn)
* (optional) [golang-migrate](https://github.com/golang-migrate/migrate)

### Clone Repo
First clone the example repo

```bash
git clone https://github.com/metalbear-co/nodejs-example.git
cd nodejs-example
```

### Setup Cluster
For this example you will need to create a few deployments and services

```bash
kubectl apply -f app.yaml
```

This command will create four deployments and relevant services
* example-pg - Postgres Database
* example-idp - [Dexidp](https://dexidp.io/) OIDC provider
* example-blog - NodeJS application located in ./blog
* example-auditor - NodeJS application located in ./auditor

After first initialization the example-idp and example-blog deployments will not work. To fix this we need to run some migrations.

### Migrations

> First establish a connection to the cluster's database in a sperate shell

Reviewer: ```suggestion
First, establish a connection to the cluster's database in a separate shell
```

```
kubectl port-forward svc/example-pg 5432:5432
```

And then run the migration script

```bash
./run_migration.sh
```

## Traffic Mirroring

A key feature of mirrord is mirroring requests from a Kubernetes container to a local process for debugging purposes.

For this example, you will first need to build the local auditor service

```bash
yarn && yarn workspace auditor build
```

Next you'll need to get the pod name from your cluster

```bash
kubectl get po -l app=example-auditor
```

The result should be something like

```
NAME                            READY   STATUS    RESTARTS   AGE
example-auditor-<hash>-<hash>   1/1     Running   0          5m
```

Now lets debug the container in the cluster by executing

```bash
mirrord exec -p example-auditor-<hash>-<hash> yarn -- workspace auditor start
```

> In a separate terminal establish a connection to the blog container by running

Reviewer: ```suggestion
In a separate terminal establish a connection to the remote blog container by running
```

Let's make a small change to `./blog/pages/index.tsx` that should print the raw response from the database when accessing the root page of the blog

```diff
export const getServerSideProps: GetServerSideProps = async ({ req }) => {
  const authorised = req.cookies['oauth-access-token'] ? !!await jwtVerify(req.cookies['oauth-access-token'], await getRemoteJWKSet()) : false;
 
-  const { rows } = await pg.query(listBlogPreviews.compile({ authorised }));
+  const result = await pg.query(listBlogPreviews.compile({ authorised }));
+
+  console.log(result);
 
   return {
     props: {
-      posts: rows
+      posts: result.rows
     },
   }
 }
```

Now lets see the output

First things first you need to get the pod name from your local cluster

```bash
kubectl get po -l app=example-blog
```

The result should be something like

```
NAME                            READY   STATUS    RESTARTS   AGE
example-blog-<hash>-<hash>      1/1     Running   0          5m
```

> \<hash\> should be the deployment and replica-set hashes generated on your cluster

And now lets test it.

```bash
mirrord exec --no-fs -x NODE_ENV -p example-blog-<hash>-<hash> yarn -- workspace blog dev
```

> > Note connection to the blog container is still required for this step, please reference [Mirroring](#traffic-mirroring)

Reviewer: ```suggestion
> Note connection to the remote blog container via port-forward is still required for this step, please reference [Mirroring](#traffic-mirroring)
```

      tableID: 16407,
      columnID: 2,
      dataTypeID: 25,
      dataTypeSize: -1,
      dataTypeModifier: -1,
      format: 'text'
    }
  ],
  _parsers: [ [Function: parseBigInteger], [Function: noParse] ],
  _types: TypeOverrides {
    _types: {
      getTypeParser: [Function: getTypeParser],
      setTypeParser: [Function: setTypeParser],
      arrayParser: [Object],
      builtins: [Object]
    },
    text: {},
    binary: {}
  },
  RowCtor: null,
  rowAsArray: false
}
```

Lets break down the command

| mirrord exec | --no-fs | -x NODE_ENV | -p example-blog-\<hash\>-\<hash\> | yarn | -- workspace blog dev |
|---|---|---|---|---|---|
||disable fs*|exclude NODE_ENV enviroment variable**|specify the running pod to mirror|executable|executable args|

\* need to disable remote fs access because of yarn caching</br>
\*\* NODE_ENV is used in webpack for compilation and it can cause conflict when running the app with dev command

## Traffic Steal

Traffic Steal lets you handle and respond to incoming requests to the pod from your local machine. In contrast to Mirroring, with Steal the response to the incoming request is sent by the local process rather than the remote pod.

> First thing we need to establish a connection to the idp so run in a sperate terminal the following command

Reviewer: ```suggestion
First thing we need to establish a connection to the remote idp. In a separate terminal, run the following command
```

```bash
kubectl port-forward svc/example-idp 5556:5556
```

Now login into the blog by navigating to [http://localhost:8080/api/login](http://localhost:8080/api/login) with the following credentials

```
email: admin@example.com
pass: password
```

> You should see a third blogpost return after the authentication is finished but if you click on it you will get a `Page Not Found`, lets fix it

Reviewer: ```suggestion
You should see a third blogpost return after the authentication is finished but if you click on it you will get a `Page Not Found` message. Lets fix it
```

 import Head from 'next/head'
@@ -6,6 +7,7 @@ import Link from 'next/link'
 
 import styles from '../../styles/Home.module.css'
 
+import { getRemoteJWKSet } from '../../lib/oidc';
 import pg from '../../lib/pg';
 
 import { findBlogPreviews } from '../../query/blog';
@@ -45,9 +47,11 @@ const Home: NextPage<{ post?: BlogPost }> = ({ post }) => {
 
 
 export const getServerSideProps: GetServerSideProps = async (context) => {
+  const authorised = context.req.cookies['oauth-access-token'] ? !!await jwtVerify(context.req.cookies['oauth-access-token'], await getRemoteJWKSet()) : false;
+
   let id = (Array.isArray(context.params?.id) ? context.params?.id[0] : context.params?.id) ?? '0';
 
-  const { rows } = await pg.query<QueryResult<typeof findBlogPreviews>>(findBlogPreviews.compile({ id }));
+  const { rows } = await pg.query<QueryResult<typeof findBlogPreviews>>(findBlogPreviews.compile({ authorised, id }));
 
   return {
     props: {
```

And lets see the results by simply adding `--steal` to the previous command

```bash
mirrord exec --no-fs -x NODE_ENV -p example-blog-<hash>-<hash> --steal yarn -- workspace blog dev
```

> And now refresh the page and the result should return now.

Reviewer: ```suggestion
Refresh the page, and you should now be able to see the result.
```

> Note connection to the blog container is still required for this step, please reference [Mirroring](#traffic-mirroring)

## Summary

mirrord can be used to either mirror incoming traffic without influencing the return values or it can be used to "override" the container, both options with full access to other resources in the cluster I.E databases other services or even mounted filesystem/configmap/secret.<br/>
This allows you to debug your process as if it was already deployed as part of the cluster.

> ### Teardown

Reviewer: Some inconsistencies here with newlines under headings
